### PR TITLE
[ARCTIC-5] Support MOR with Flink Engine in runtime batch mode.

### PR DESCRIPTION
**Pull request by YesOrNo828**
## Why are the changes needed?

Add support for MOR with Flink Engine.
Fixed: #5 

## Brief change log

- Add MergeOnReadIncrementalPlanner to implement the ContinuousSplitPlanner interface;
- Add FlinkArcticMORDataReader to adapt the AbstractAdaptHiveArcticDataReader;
- Add MixedIncrementalLoader: a mixed-format table(mixed iceberg, mixed-hive) incremental loader.

## How was this patch tested?
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: core/src/main/java/com/netease/arctic/scan/TableEntriesScan.java [hunk 1]
```diff
@@ -144,6 +146,11 @@ public Builder includeColumnStats() {
       return this;
     }
 
+    public Builder fromSequence(Long fromSequence) {
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/scan/TableEntriesScan.java:149]: We may need to add a unit test for the new method.

File: flink/v1.14/flink/src/main/java/com/netease/arctic/flink/read/FlinkSplitPlanner.java [hunk 2]
```diff
@@ -60,6 +68,17 @@ public static List<ArcticSplit> planFullTable(KeyedTable keyedTable, AtomicInteg
     return planFullTable(baseAndChangeTask, splitCount);
   }
 
+  public static List<ArcticSplit> planFullTable(
```
Comment (zhoujinsong) [inline flink/v1.14/flink/src/main/java/com/netease/arctic/flink/read/FlinkSplitPlanner.java:71]: I suggest adding some Javadoc here to tell the function of the method.

File: flink/v1.14/flink/src/main/java/com/netease/arctic/flink/read/FlinkSplitPlanner.java [hunk 3]
```diff
@@ -75,6 +94,27 @@ private static List<ArcticSplit> planFullTable(
     return allSplits;
   }
 
+  public static List<ArcticSplit> mergeOnReadPlan(
```
Comment (zhoujinsong) [inline flink/v1.14/flink/src/main/java/com/netease/arctic/flink/read/FlinkSplitPlanner.java:97]: I suggest adding some Javadoc here to tell the function of the method, especially the difference with `planFullTable`.

File: flink/v1.14/flink/src/main/java/com/netease/arctic/flink/read/MixedIncrementalLoader.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+      LOG.info("Currently, queue contain {} splits, scan position is {}.", splitQueue.size(), enumeratorPosition.get());
+      return !splitQueue.isEmpty();
+    }
+    return true;
+  }
+
+  public CloseableIterator<T> next() {
+    ArcticSplit split = splitQueue.poll();
+    if (split == null) {
+      throw new IllegalArgumentException("next() called, but no more valid splits");
```
Comment (zhoujinsong) [inline flink/v1.14/flink/src/main/java/com/netease/arctic/flink/read/MixedIncrementalLoader.java:93]: Is `IllegalStateExcpetion` more appropriate here？

File: flink/v1.14/flink/src/main/java/com/netease/arctic/flink/read/hybrid/enumerator/ContinuousSplitPlanner.java [hunk 5]
```diff
@@ -33,4 +35,10 @@ public interface ContinuousSplitPlanner extends Closeable {
    * Discover the files appended between {@code lastPosition} and current table snapshot
    */
   ContinuousEnumerationResult planSplits(ArcticEnumeratorOffset lastPosition);
```
Comment (zhoujinsong) [inline flink/v1.14/flink/src/main/java/com/netease/arctic/flink/read/hybrid/enumerator/ContinuousSplitPlanner.java:37]: Maybe we can add a default implementation for this method like :
```
return planSplits(lastPosition, Lists.newArrayList());
```

File: flink/v1.14/flink/src/main/java/com/netease/arctic/flink/read/hybrid/enumerator/ContinuousSplitPlannerImpl.java [hunk 6]
```diff
@@ -72,16 +73,61 @@ public ContinuousEnumerationResult planSplits(ArcticEnumeratorOffset lastOffset)
...
+          planChangeTable(entriesScan, fromSequence, table.changeTable().spec(), splitCount);
+      return new ContinuousEnumerationResult(
+          arcticChangeSplit,
+          lastPosition,
+          ArcticEnumeratorOffset.of(snapshotId, null));
+    }
+    return ContinuousEnumerationResult.EMPTY;
+  }
+
+  protected ContinuousEnumerationResult discoverIncrementalSplits(
```
Comment (zhoujinsong) [inline flink/v1.14/flink/src/main/java/com/netease/arctic/flink/read/hybrid/enumerator/ContinuousSplitPlannerImpl.java:116]: These two `discoverIncrementalSplits` have so much identical code, perhaps we can merge them.

File: flink/v1.14/flink/src/main/java/com/netease/arctic/flink/read/hybrid/enumerator/ContinuousSplitPlannerImpl.java [hunk 7]
```diff
@@ -113,4 +159,20 @@ private ContinuousEnumerationResult discoverInitialSplits() {
         null,
         ArcticEnumeratorOffset.of(changeStartSnapshotId, null));
   }
+
+  protected ContinuousEnumerationResult discoverInitialSplits(List<Expression> filters) {
```
Comment (zhoujinsong) [inline flink/v1.14/flink/src/main/java/com/netease/arctic/flink/read/hybrid/enumerator/ContinuousSplitPlannerImpl.java:163]: These two `discoverInitialSplits` have so much identical code, perhaps we can merge them.

File: flink/v1.14/flink/src/main/java/com/netease/arctic/flink/read/hybrid/enumerator/MergeOnReadIncrementalPlanner.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    return new ContinuousEnumerationResult(
+        arcticSplits,
+        null,
+        ArcticEnumeratorOffset.of(changeStartSnapshotId, null));
+  }
+
+  @Override
+  public void close() throws IOException {
```
Comment (zhoujinsong) [inline flink/v1.14/flink/src/main/java/com/netease/arctic/flink/read/hybrid/enumerator/MergeOnReadIncrementalPlanner.java:66]: The override `close` method seems have the same implementation with the super class.

**Review by zhoujinsong — COMMENTED** [review]
@YesOrNo828 Is this still a WIP work as I have noticed that the affected Flink version is just 1.14 and did not find the SQL implementation?

**Comment by YesOrNo828** [comment]
> 

@zhoujinsong  Thanks for reviewing, I have shipped the code to the flink1.12 and 1.15 modules. 

This pr only exposes the reader API by merge-on-read mode and excludes the SQL implementation. I'll open a new issue to follow the SQL implementation for MOR.

**Review by zhoujinsong — APPROVED** [review]
LGTM.

**Pull request merged**